pymongo/pool.py

```diff
@@ -1289,6 +1335,15 @@ def get_socket(self, all_credentials, checkout=False):
             if not checkout:
                 self.return_socket(sock_info)
 
+    def _raise_if_not_ready(self):
+        if self.opts.pause_enabled and self.state == PAUSED:
+            if self.enabled_for_cmap:
+                self.opts.event_listeners.publish_connection_check_out_failed(
+                    self.address, ConnectionCheckOutFailedReason.CONN_ERROR)
+            # TODO: ensure this error is retryable
```

Comment:
@patrickfreed have you thought about this question yet? How will we test that this error is retryable in all drivers?

Reply:
I left a comment discussing a few of the options on the SPEC PR: https://github.com/mongodb/specifications/pull/878#discussion_r527877646

As right now, I'd like to wait until retryable reads gets the unified format tests before personally writing any of them for the spec so we can get this project closed out, though I could add a prose test to it though if you think we should.

Something like:
1. Client with maxPoolSize=1
2. failCommand insert with blockTimeMS=1000, errorCode: \<something retryable that clears the pool\> 
3. insert in thread1
4. insert in thread2
5. ensure both inserts succeed

Reply:
Can you open a new DRIVERS ticket to add this test?

Reply:
Opened https://jira.mongodb.org/browse/DRIVERS-1483, though it made me realize that errors during connection establishment may or may not be retryable. The spec is mostly silent on it. I think we definitely want this error to be retryable though, since no attempt at the command is actually made. I think given this confusion, we'll need to sort out this test before we close out DRIVERS-781.

Reply:
So it seems to hinge on https://jira.mongodb.org/browse/DRIVERS-746, which may be a larger effort than we'd want to drag into DRIVERS-781. We could consider _not_ testing this behavior since we're technically not making anything worse than it already is (those threads in the WaitQueue will likely fail to make their connections anyways). Ideally we would complete DRIVERS-746 and then could have this retry though.

Reply:
I added a test for this in python. Pymongo already retries after (retryable) connection handshake errors so the test passes as expected.